### ISSUE TITLE
[autocast scaler] unable to prevent lr scheduler before optimizer step warning

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#55585]
## 🐛 Bug

We get:

```
torch/optim/lr_scheduler.py:132: UserWarning: Detected call of lr_scheduler.step() before optimizer.step(). In PyTorch 1.1.0 and 
later, you should call them in the opposite order: optimizer.step() before lr_scheduler.step(). Failure to do this will result in PyTorch 
skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
warnings.warn("Detected call of lr_scheduler.step() before optimizer.step().
```

because autocast scaler skips `optimizer.step` on overflow.

## To Reproduce

```
        scaler = torch.cuda.amp.GradScaler()
        with torch.cuda.amp.autocast():
            output = net(input)
            loss = loss_fn(output, target)
            scaler.scale(loss).backward()
            scaler.unscale_(optimizer)
            scaler.step(optimizer)
            scaler.update()    
            scheduler.step()
```

Under `autocast` if scaler sees an overflow the first time `scaler.step(optimizer)` is called, it won't call `optimizer.step`, so as you can see we have a problem here, ` scaler.step(optimizer)` doesn't return a status of whether it was run or not and therefore it's impossible to know whether to run `scheduler.step()` or not.

So the warning is correct, but there is no programmatic way to fix it.
It'd be nice to be able to get rid of this warning by solving this situation.

Original report/discussion: https://github.com/huggingface/transformers/issues/11106

Thank you!

**Comment by ptrblck** [comment]
> So the warning is correct, but there is no programmatic way to fix it.

As a workaround until a non-synchronizing approach is found, you could check, if the scale value was decreased and skip the `scheduler.step()` in this case as given [here](https://discuss.pytorch.org/t/model-weights-not-getting-updated-when-using-autocast/117286/10?u=ptrblck).

**Comment by stas00** [comment]
Awesome! That was super helpful, @ptrblck!

So this works:
```
        scaler = torch.cuda.amp.GradScaler()
        with torch.cuda.amp.autocast():
            output = net(input)
            loss = loss_fn(output, target)
            scaler.scale(loss).backward()
            scaler.unscale_(optimizer)
            scale_before = self.scaler.get_scale()
            scaler.step(optimizer)
            scaler.update()    
            scale_after = self.scaler.get_scale()
            if scale_before == scale_after:
                scheduler.step()
```

I don't know if this should be closed or remain for the integral solution.
